custom-recipes/pi-system-retrieve-list/recipe.py

```diff
@@ -75,6 +76,16 @@
                 is_debug_mode=is_debug_mode, network_timer=network_timer
             )
             previous_server_url = server_url
+            if time_not_parsed:
+                # make sure all OSIsoft time string format are evaluated at the same time
+                # rather than at every request, at least for start / end times set in the UI
+                time_not_parsed = False
+                if not use_start_time_column:
+                    start_time = client.parse_pi_time(start_time)
+                if not use_end_time_column:
+                    end_time = client.parse_pi_time(end_time)
+                sync_time = client.parse_pi_time(sync_time)
+
```

Comment:
I meant to do it once at the start actually. Now I see why you do it this way - we don't have an osisoft client yet and we need one to do the conversion (and the server URL can depend on the row in the input dataset). However this implies there might be another more subtle problem. If the OSI server URL changes between rows, the clocks on these servers might be different, so their interpretation of relative datetimes might be different. It would be best to use the server we are targeting to do the conversion. So ideally we should reparse each time the server URL changes.

Which means taking out the if statement:
`if time_not_parsed:`

And also retaining the times that were input via the config and using a different variable for the parsed times.

Not sure it would be a big problem but would probably be best to avoid it.

Reply:
We discussed this and we should leave as it is - reparsing for rows might bring back the original problem. 
The first server encountered can be used for the reference time, we will do that soon after the action is triggered.
If the OSI server clocks are in sync there should not be a problem - we can add something to the docs about this.